### PR TITLE
[Fix] Validation and select box

### PR DESCRIPTION
**Pull request by yorkittran**
# Validation
For some integer fields like year of manufacturer, memory amount, quantity, price

# Select box
Change input text with autocomplete into select box when manager trying to create a new inventory. Beside, combine 3 input of OS version into 1 input text and validate with it.
__1 _3000_en_memories

### CODE REVIEW

**Comment by raksul-sian-izm** [comment]
This time, I don’t care about code’s readability very much, but please fix to return understandable error message!

**Comment by yorkittran** [comment]
I fix the error message, it looks like this
__1 _3000_en_inventories

File: app/controllers/inventories_controller.rb [hunk 1]
```diff
@@ -1,37 +1,29 @@
...
   before_action :authenticate_user!
-  before_action :set_model, only: %i[show destroy restore]
+  before_action :set_model, only: %i[destroy restore]
+  before_action :set_data, only: %i[new create]
 
   # GET /inventories
   # GET /inventories.json
   def index
-    @warehouse = Phone.joins(:inventories, :memory, :model).select(:model_id, :name, 'SUM(quantity) as total_quantity', 'MAX(price) as max_price', 'MIN(price) as min_price', 'MAX(amount) as max_amount', 'MIN(amount) as min_amount').group(:model_id) # rubocop:disable Layout/LineLength
+    @warehouse = Phone.joins(:memory).joins("INNER JOIN inventories ON inventories.phone_id = phones.id AND inventories.discarded_at IS NULL").joins("INNER JOIN models ON models.id = phones.model_id AND models.discarded_at IS NULL").select(:model_id, :name, 'SUM(quantity) as total_quantity', 'MAX(price) as max_price', 'MIN(price) as min_price', 'MAX(amount) as max_amount', 'MIN(amount) as min_amount').group(:model_id) # rubocop:disable Layout/LineLength
```
Comment (raksul-sian-izm) [inline app/controllers/inventories_controller.rb]: I think you can write like this with`where` clause.
```ruby
Phone.joins(:memory, :inventories, :model).select(:model_id, :name, 'SUM(quantity) as total_quantity', 'MAX(price) as max_price', 'MIN(price) as min_price', 'MAX(amount) as max_amount', 'MIN(amount) as min_amount').where('models.discarded_at IS NULL').where('inventories.discarded_at IS NULL').group(:model_id)
```

File: app/controllers/inventories_controller.rb [hunk 2]
```diff
@@ -1,37 +1,29 @@
...
-  end
+  def new; end
 
   # POST /inventories
   # POST /inventories.json
   def create
     respond_to do |format|
-      if CreateBulkService.call(convert_params, current_user)
+      begin
+        CreateBulkService.call(convert_params, current_user)
```
Comment (raksul-sian-izm) [inline app/controllers/inventories_controller.rb]: [IMO]
Name the target record name in case other class's record should be created. For example, `Inventory::BulkCreateService`

File: app/controllers/inventories_controller.rb (hunk 1 above)
Comment (raksul-sian-izm) [reply app/controllers/inventories_controller.rb]: Like this complex query makes code unreadable. Even though it's hard for senior members to maintain.
So, another development policy is like this.
In this architecture, you can keep controller simple. Though the number of query may be N+1, it won't be big performance problem as long as introducing paging.
```ruby
# controller
@models = Model.includes(:phones).page.per

# view file
@models.each |model|
  model.phones.count # count information
  model.max_price 
...
end

# model
class Model
  def max_price
    phones.order(:price).first
  end
end

```

Of course, you need experience to be able to do this decision in real situation.
Then, in this training contents, or at the beginning, please keep in mind 'Keep as simple and small as possible'

File: app/controllers/inventories_controller.rb [hunk 3]
```diff
@@ -67,18 +59,42 @@ def model_params
...
-      item[:model] = Model.where(name: value.split(' (')[0]).first
-      item[:body_color] = BodyColor.where(name: params[:body_color][:name][key]).first_or_initialize
-      item[:memory] = Memory.where(amount: params[:memory][:amount][key], display_name: params[:memory][:display_name][key]).first_or_initialize
-      item[:os_version] = OsVersion.where(major: params[:os_version][:major][key], minor: params[:os_version][:minor][key], patch: params[:os_version][:patch][key]).first_or_initialize
-      item[:inventory] = Inventory.new(quantity: params[:inventory][:quantity][key], price: params[:inventory][:price][key])
+      item = {}
+      item[:model] = Model.find(value)
+      item[:body_color] = BodyColor.find(params[:body_color][key])
+      item[:memory] = Memory.find(params[:body_color][key])
+      item[:os_version] = OsVersion.where(
```
Comment (raksul-sian-izm) [inline app/controllers/inventories_controller.rb:79]: This three types os_version's validation and initialization are really complicating.
If I developed in real situation, 

- Define `os_version` in varchar
- I communicate PO that splitting os_version on interface is unnecessary for real operation.
- However, still if PO asked me to split os_version to three part as you developed now, I would arrange front end part. Frontend concats the three values into one value and post this endpoint with the one value.
- Server side validates only that version is empty or not. 

This process will be really simpler, too.
https://github.com/yorkittran/raksul-project-training/pull/7/files?file-filters%5B%5D=.js&file-filters%5B%5D=.rb&file-filters%5B%5D=.slim&hide-deleted-files=true#diff-110e8bbf56e45581ff6d5f5f86fb06eeR90

File: app/controllers/inventories_controller.rb [hunk 4]
```diff
@@ -67,18 +59,42 @@ def model_params
...
+        major: os_version[0],
+        minor: os_version[1].nil? ? 0 : os_version[1],
+        patch: os_version[2].nil? ? 0 : os_version[2],
```
Comment (raksul-sian-izm) [inline app/controllers/inventories_controller.rb:82]: Default value for any table's column should be common for most of the case.
Then, it should be defined in model class. like this way.
https://stackoverflow.com/a/4576026

File: app/javascript/packs/card.js [hunk 5]
```diff
@@ -1,47 +1,42 @@
+let modelSelected = document.getElementById("model-select");
```
Comment (raksul-sian-izm) [inline app/javascript/packs/card.js:1]: Nowadays, ordinally there is no case to write many Javascript without any frontend framework like Vue, React... 
Because it's really hard to keep simplicity, readability and maintainability of the code.

In real development, you will face the situation that PM ask you to superior UI. In that case, please stop and consider "Shouldn't it be developed by frontend?, or is it really necessary interface?" 

There are two different types of applications in Raksul.
**Web** for public user and **Ops** for internal operators.

For public user's application, we're often required to maximize the user interface. Then, like this Javascript front interface are necessary and it's maintained with Vue.js.
On the other hand, introducing the frontend framework requires learning cost, too. Then, for OPS, we develop the interface with marginal usability and many time software engineer should communicate PM to minimize the development cost and take much more development time for public user's application.

In conclusion, if you will develop like this operator's form, you may want to develop simpler even if the interface isn't fashionate.

Comment (raksul-sian-izm) [reply app/javascript/packs/card.js:1]: If you create form simpler, 

- Registering model and model's information should be same form without Javascript very much
- At first scope, we don't support multiple registration but one by one.
- If shop manager ask to register multiple models at the same time, develop to import csv data.

**Comment by raksul-sian-izm** [comment]
You don't need to refine this time because it'll be huge refine. Then, please read my comment. If you have any question, feel free to ask me! 
With understanding my comments, adopting the knowledge for real project is better!

**Comment by raksul-sian-izm** [comment]
The important point is 
- Our products are expected to be used for long time, more than few years. 
- They're also refined many time for improving user impact. Then, simple and maintainable code are really important.

File: app/controllers/inventories_controller.rb (hunk 1 above)
Comment (yorkittran) [reply app/controllers/inventories_controller.rb]: Actually I chose the way using this query to avoid N+1, cause I'm using [datatable](https://datatables.net/) to perform pagination with JavaScript (not with kaminari gem), including search and sort. It required all the record I want to show up at once, so if I use your solution, I may change quite big (using pagination at server side, re-designing table on view, using search with table). At first, I thought about your solution but looking the pros of datatable made me chose this.

File: app/controllers/inventories_controller.rb (hunk 2 above)
Comment (yorkittran) [reply app/controllers/inventories_controller.rb]: I got this

File: app/javascript/packs/card.js (hunk 5 above)
Comment (yorkittran) [reply app/javascript/packs/card.js:1]: About technical, I understand how hard to read my JS code and my form will be much more simpler if manager register inventory one by one (validation and creating may be simpler too).

But after the discussion with Nhan san during step 1 (wireframes), I think register multiple inventories is not a bad idea at all. When I asked Nhan san about import CSV data to register multiple inventories, he told me to think about is it really necessary for manager to make a CSV file with a few records. IMO, this application is make for used phone shops and it is not a big number of phone's types when the shop owner (manager) import these phone daily, weekly or monthly.

So anyway, registration for multiple inventories at once makes me interested and I tried to do this (cause I'm not familiar with FE framework so JS is the comfortable way that help me create this feature). It must be easier if I do it one by one or using CSV to import an amount of phones.

File: app/controllers/inventories_controller.rb (hunk 3 above)
Comment (yorkittran) [reply app/controllers/inventories_controller.rb:79]: This solution helps me a lot! So can I remove the os_version table and add a column os_version into phones table? It can be much more simpler.

File: app/controllers/inventories_controller.rb (hunk 4 above)
Comment (yorkittran) [reply app/controllers/inventories_controller.rb:82]: In this case, I'm using select query with where so I'm not sure about understanding your comment.
BTW, about default value, I set 0 is a default value of minor and patch in database (through migration files). Is it ok or we should move the default value in model?

File: app/controllers/inventories_controller.rb [hunk 6]
```diff
@@ -1,37 +1,29 @@
...
+      begin
+        Inventories::BulkCreateService.call(convert_params, current_user)
         format.html { redirect_to inventories_path, notice: 'Inventory was successfully imported.' }
         format.json { render :index, status: :created, location: inventories_path }
-      else
+      rescue ActiveRecord::RecordInvalid => e
+        flash.now[:alert] = e.message
         format.html { render :new }
-        format.json { render json: 'error', status: :unprocessable_entity }
+        format.json { render json: @inventory, status: :unprocessable_entity }
```
Comment (raksul-sian-izm) [inline app/controllers/inventories_controller.rb:26]: It looks weird to return `@inventory` object in spite process is failing 🤔 
Usually when error happen, only error message should be returned.

File: app/controllers/inventories_controller.rb (hunk 4 above)
Comment (raksul-sian-izm) [reply app/controllers/inventories_controller.rb:82]: > I set 0 is a default value of minor and patch in database (through migration files)

I usually define only database's default value or both of database and application.
Defining database's default value is more certain for keeping data consistency because when application is transferred another application but database keep being used, the default value is also same.

When the object is only initialized, database's default value isn't set. Then, defining default value on application is required.

File: app/javascript/packs/card.js (hunk 5 above)
Comment (raksul-sian-izm) [reply app/javascript/packs/card.js:1]: That makes sense!
So, in real situation, let's communicate Product Manager about the user's requirement well.

**Review by raksul-sian-izm — COMMENTED** [review]
> Actually I chose the way using this query to avoid N+1, cause I'm using datatable to perform pagination with JavaScript (not with kaminari gem), including search and sort. It required all the record I want to show up at once, so if I use your solution, I may change quite big (using pagination at server side, re-designing table on view, using search with table). At first, I thought about your solution but looking the pros of datatable made me chose this.

As long as small number of record, current architecture won't be problem.
However, if the number of models and joining tables records will increase, the current architecture may be problem. Whereas, performance with pagination on server side can keep the performance level well even though the number of records are increased.
For this used-smartphone's maintenance application, your architecture won't be big problem for long term operation.

**Comment by raksul-sian-izm** [comment]
LGTM 🍭
